opal/threads/thread_usage.h

```diff
@@ -0,0 +1,213 @@
+#define OPAL_THREAD_USAGE_H
+
+#include "opal/sys/atomic.h"
+#include "opal_config.h"
+
+/**
+ * Check and see if the process is using multiple threads.
+ *
+ * @retval true If the process may have more than one thread.
+ * @retval false If the process only has a single thread.
+ *
+ * The value that this function returns is influenced by:
+ *
+ * - how MPI_INIT or MPI_INIT_THREAD was invoked,
+ * - what the final MPI thread level was determined to be,
+ * - whether the OMPI or MPI libraries are multi-threaded
+ *
+ * MPI_INIT and MPI_INIT_THREAD (specifically, back-end OMPI startup
+ * functions) invoke opal_set_using_threads() to influence the value of
+ * this function, depending on their situation. Some examples:
+ *
+ * - if MPI_INIT is invoked, and the ompi components in use are
+ * single-threaded, this value will be false.
+ *
+ * - if MPI_INIT_THREAD is invoked with MPI_THREAD_MULTIPLE, we have
+ * thread support, and the final thread level is determined to be
+ * MPI_THREAD_MULTIPLE, this value will be true.
+ *
+ * - if the process is a single-threaded OMPI executable (e.g., mpicc),
+ * this value will be false.
+ *
+ * Hence, this function will return false if there is guaranteed to
+ * only be one thread in the process.  If there is even the
+ * possibility that we may have multiple threads, true will be
+ * returned.
+ */
+#if OMPI_ENABLE_THREAD_MULTIPLE
+
+OPAL_DECLSPEC extern bool opal_uses_threads;
+#define opal_using_threads()  opal_uses_threads
+
+#else
+
+#define opal_using_threads()  false
+
+#endif /* OMPI_ENABLE_THREAD_MULTIPLE */
+
+/**
+ * Set whether the process is using multiple threads or not.
+ *
+ * @param have Boolean indicating whether the process is using
+ * multiple threads or not.
+ *
+ * @retval opal_using_threads The new return value from
+ * opal_using_threads().
+ *
+ * This function is used to influence the return value of
+ * opal_using_threads().  If configure detected that we have thread
+ * support, the return value of future invocations of
+ * opal_using_threads() will be the parameter's value.  If configure
+ * detected that we have no thread support, then the retuen from
+ * opal_using_threads() will always be false.
+ */
+static inline bool opal_set_using_threads(bool have)
+{
+#if OMPI_ENABLE_THREAD_MULTIPLE
+    opal_uses_threads = have;
+#else
+    have = true;               /* just shut up the compiler */
+#endif
+    return opal_using_threads();
+}
+
+
+/**
+ * Use an atomic operation for increment/decrement if opal_using_threads()
+ * indicates that threads are in use by the application or library.
+ */
+
+#if OMPI_ENABLE_THREAD_MULTIPLE
+static inline int32_t
+OPAL_THREAD_ADD32(volatile int32_t *addr, int delta)
+{
+    int32_t ret;
+
+    if (opal_using_threads()) {
+        ret = opal_atomic_add_32(addr, delta);
+    } else {
+        ret = (*addr += delta);
+    }
+
+    return ret;
+}
+#else
+static inline int32_t
+OPAL_THREAD_ADD32(volatile int32_t *addr, int delta)
```

Comment:
@hjelmn do we need two versions of OPAL_THREAD_ADD32, OPAL_THREAD_ADD64 and OPAL_THREAD_ADD_SIZE_T. Wouldn't the compiler remove the dead code if opal_using_threads() returns false?